mwapi/async_session.py

```diff
@@ -0,0 +1,213 @@
+            no trailing "/".
+        user_agent : `str`
+            The User-Agent header to include with all requests.  Use this field
+            to identify your script/bot/application to system admins of the
+            MediaWiki API you are using.
+        formatversion : int
+            The formatversion to supply to the API for all requests.
+        api_path : `str`
+            The path to "api.php" on the server -- must begin with "/".
+        timeout : `float`
+            How long to wait for the server to send data before giving up
+            and raising an error (
+            :class:`aiohttp.client_exceptions.ServerTimeoutError` or
+            :class:`asyncio.exceptions.TimeoutError`).
+            By default aiohttp uses a total 300 seconds (5min) timeout.
+        session : `aiohttp.ClientSession`
+            (optional) an `aiohttp` session object to use
+    """
+
+    def __init__(self, host, user_agent=None, formatversion=None,
+                 api_path=None,
+                 timeout=None, session=None, **session_params):
+        self.host = str(host)
+        self.formatversion = int(formatversion) \
+            if formatversion is not None else None
+        self.api_path = str(api_path or "/w/api.php")
+        self.api_url = self.host + self.api_path
+        self.timeout = float(timeout) if timeout is not None else None
+        self.session = session or aiohttp.ClientSession()
+        for key, value in session_params.items():
+            setattr(self.session, key, value)
+
+        self.headers = {}
+
+        if user_agent is None:
+            logger.warning("Sending requests with default User-Agent.  " +
+                           "Set 'user_agent' on mwapi.Session to quiet this " +
+                           "message.")
+            self.headers['User-Agent'] = DEFAULT_USERAGENT
+        else:
+            self.headers['User-Agent'] = user_agent
+
+    async def _request(self, method, params=None, auth=None):
+        params = params or {}
+        if self.formatversion is not None:
+            params['formatversion'] = self.formatversion
+
+        if method.lower() == "post":
+            data = params
+            data['format'] = "json"
+            params = None
+
+        else:
+            data = None
+            params = params or {}
+            params['format'] = "json"
+
+        try:
+            async with self.session.request(method=method, url=self.api_url,
+                                            params=params, data=data,
+                                            timeout=self.timeout,
+                                            headers=self.headers,
+                                            verify_ssl=True,
+                                            auth=auth) as resp:
+
+                try:
+                    doc = await resp.json()
+                except (ValueError, aiohttp.ContentTypeError):
```

Comment:
Could this except be placed with the other ones below? afaics if there is a ValueError or ContentTypeError we raise an exception anyway (at line 94) so we don't execute the code afterwards. Does it make sense?

Reply:
Yes, it makes sense! Done.